### PR TITLE
Deregistration UI websocket events

### PR DESCRIPTION
**Pull request by rtorrero**
# Description
This PR adds the code to handle the state changes in react upon the arrival of the `*_deregistered` events from the host, cluster and sap_system aggregates.

Remember to add the right labels to this PR.

## How was this tested?
Manually, issuing Host deregistration requests in the REPL. End-to-end tests will be added once the UI changes are completed.

### CODE REVIEW

**Review by arbulu89 — COMMENTED** [review]
@rtorrero I think this is a good chance to start adding jest unit tests to this changes

File: assets/js/state/clusters.test.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import clustersReducer, { removeCluster } from '@state/clusters';
+
+describe('Clusters reducer', () => {
+  it('should remove cluster from state', () => {
+    const cluster = { id: 'test-cluster-id' };
```
Comment (arbulu89) [inline assets/js/state/clusters.test.js:5]: Most probably it would be better to use the `clusterFactory`.
```
const { id } = clusterFactory.build()
# or
const clusters = cluster.Factory.buildList(5) # with some further work, which would do a more robust test, just removing 1 from the list, instead of looking for an empty list
```

The same goes for the other tests

File: assets/js/state/sapSystems.js [hunk 2]
```diff
@@ -100,6 +100,11 @@ export const sapSystemsListSlice = createSlice({
         return sapSystem;
       });
     },
+    removeSAPSystem: (state, { payload: { id } }) => {
+      state.sapSystems = state.sapSystems.filter(
+        (sapSystem) => id !== sapSystem.id
```
Comment (arbulu89) [inline assets/js/state/sapSystems.js:105]: Following the others you would do `sapSystem.id !== id`

File: assets/js/state/sagas/clusters.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  yield put(removeCluster(payload));
+  yield put(
+    appendEntryToLiveFeed({
+      source: payload.name,
+      message: 'Cluster deregistered.',
+    })
+  );
+  yield put(
+    notify({
+      text: `The cluster ${payload.name} has been deregistered.`,
```
Comment (arbulu89) [inline assets/js/state/sagas/clusters.js:16]: Keep in mind that not all clusters have a name, so this could end up empty.
Maybe something like `${payload.name || payload.id}` ?

File: assets/js/state/sagas/clusters.test.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { recordSaga } from '@lib/test-utils';
+import { clusterDeregistered } from '@state/sagas/clusters';
+import { removeCluster } from '@state/clusters';
+
+describe('Clusters sagas', () => {
+  it('should trigger reducer to remove cluster', async () => {
```
Comment (arbulu89) [inline assets/js/state/sagas/clusters.test.js:6]: Maybe `should remove the cluster` simply?

File: assets/js/state/sagas/clusters.test.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    const payload = {
```
Comment (arbulu89) [inline assets/js/state/sagas/clusters.test.js:7]: Use factories

File: assets/js/state/sagas/index.js [hunk 6]
```diff
@@ -205,6 +208,10 @@ function* watchHeartbeatFailed() {
   yield takeEvery('HEARTBEAT_FAILED', heartbeatFailed);
 }
 
+function* watchHostDeregistered() {
+  yield takeEvery('HOST_DEREGISTERED', hostDeregistered);
```
Comment (arbulu89) [inline assets/js/state/sagas/index.js:212]: Let's move this `watch` to the saga files (as we do with the catalog and lastExecutions).
Besides that, `HOST_DEREGISTERED` should be a constant.
By now, these are stored in the `actions` file, but with the latest ADR i'm not sure where this should go. Reducer or saga? @dottorblaster wdyt?

File: lib/trento/application/projectors/sap_system_projector.ex [hunk 7]
```diff
@@ -207,11 +209,14 @@ defmodule Trento.SapSystemProjector do
         _,
         _
       ) do
+    %SapSystemReadModel{sid: sid} = Repo.get!(SapSystemReadModel, sap_system_id)
```
Comment (arbulu89) [inline lib/trento/application/projectors/sap_system_projector.ex:212]: Doesn't this raise some any failure in tests?
We should have a look, or at least try to include this change in the tests. 
Maybe the assertion looks for partial objects

**Review by arbulu89 — CHANGES_REQUESTED** [review]
In general it looks good.
I think that we should use more the factories in the tests if possible.
Besides that, I commented some topics about the placement of functions and usage of constants.

Finally, What about the `databases`?

File: assets/js/state/sagas/clusters.test.js (hunk 4 above)
Comment (jamie-suse) [reply assets/js/state/sagas/clusters.test.js:6]: This test, is testing that the saga dispatches the correct action to the reducer. i.e. it does not validate the state change (because we test this in the reducer test).

Comment (arbulu89) [reply assets/js/state/sagas/clusters.test.js:8]: But that's going technical...
At the end this saga aims to update our state internally, in our domain terminology this is to `remove a cluster`.
Anyway, I don't want to force anything, it is just my point of view

Comment (jamie-suse) [reply assets/js/state/sagas/clusters.test.js]: Okay no worries, changed it. Was just making sure we are on the same page 👍

File: assets/js/state/clusters.test.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  it('should remove cluster from state', () => {
+    const cluster = clusterFactory.build();
+    const initialState = {
+      clusters: [cluster],
+    };
+
+    const action = removeCluster(cluster);
+
+    const expectedState = {
+      clusters: [],
```
Comment (arbulu89) [inline assets/js/state/clusters.test.js:14]: I still think that this test is weak. I would prefer having a removal from a list where you have more than one element

**Review by arbulu89 — COMMENTED** [review]
Hey @jamie-suse 
It looks much better.
Waiting to confirm where to put the new `action` constants.
And please, have a look on the comments I wrote about the `id/cluster_id` possible difference.
I think we would be fine to merge once that is checked

File: assets/js/state/actions/cluster.js [hunk 9]
```diff
@@ -1,6 +0,0 @@
-export const CHECKS_SELECTED = 'CHECKS_SELECTED';
```
Comment (arbulu89) [inline assets/js/state/actions/cluster.js:1]: I don't understand why you do these changes in this PR :sweat_smile: 
I would create tech-debt tasks and work on them. 
Otherwise the reviews got out of hand.
(And I know that I asked you to not use action files, but you could have only do things belonging to the actual change there)

File: assets/js/state/sagas/clusters.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { put, takeEvery } from 'redux-saga/effects';
+import { CLUSTER_DEREGISTERED, removeCluster } from '@state/clusters';
+import { appendEntryToLiveFeed } from '@state/liveFeed';
+import { notify } from '@state/actions/notifications';
+
+export function* clusterDeregistered({ payload: { name, cluster_id } }) {
+  yield put(removeCluster({ id: cluster_id }));
```
Comment (arbulu89) [inline assets/js/state/sagas/clusters.js:7]: It was not simply better to change the projection code to use `id` instead of `cluster_id`?
To avoid transformations that we might have issues finding in the future

File: assets/js/state/sagas/hosts.test.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import { recordSaga } from '@lib/test-utils';
+import { hostDeregistered } from '@state/sagas/hosts';
+import { HOST_DEREGISTERED, removeHost } from '@state/hosts';
+import { hostFactory } from '@lib/test-utils/factories';
+
+describe('Hosts sagas', () => {
+  it('should remove the host', async () => {
+    const { id, hostname } = hostFactory.build();
```
Comment (arbulu89) [inline assets/js/state/sagas/hosts.test.js:8]: This doesn't make much sense, right?
Why not simply do `const payload = hostFactory.build()` ?
Having more fields from the factory doesn't affect anything as far as I see

**Review by arbulu89 — APPROVED** [review]
@jamie-suse I'm approving this because I see it is stalling.
But I still think that there are things to be improved.

Other thing that now I miss is the instances live update.
We don't listen to any message to remove the sap app/database instances, so what happens when we remove them?
I would expect that they are removed from many views (host, sap system, database).
Could you track this missing thing and still work on it?

File: assets/js/state/sagas/hosts.test.js (hunk 11 above)
Comment (jamie-suse) [reply assets/js/state/sagas/hosts.test.js:8]: The extra fields cause the check using `toContainEqual()` to fail. Trying to see if there's an easy way to get it to work in Jest.

File: assets/js/state/sagas/clusters.test.js [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+import { clusterDeregistered } from '@state/sagas/clusters';
+import { CLUSTER_DEREGISTERED, removeCluster } from '@state/clusters';
+import { clusterFactory } from '@lib/test-utils/factories';
+
+describe('Clusters sagas', () => {
+  it('should remove the cluster', async () => {
+    const { id, name } = clusterFactory.build();
+
+    const dispatched = await recordSaga(clusterDeregistered, {
+      type: CLUSTER_DEREGISTERED,
```
Comment (CDimonaco) [inline assets/js/state/sagas/clusters.test.js:11]: We should use the value of CLUSTER_DEREGISTERED as expected value, not the const imported,

File: assets/js/state/sagas/hosts.test.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+    const payload = { id, hostname };
+
+    const dispatched = await recordSaga(hostDeregistered, {
+      type: HOST_DEREGISTERED,
```
Comment (CDimonaco) [inline assets/js/state/sagas/hosts.test.js:12]: Why we need the type field here>
I think we don't need the type value here, we could just easily add the payload like the other tests

**Review by CDimonaco — CHANGES_REQUESTED** [review]

File: assets/js/state/sagas/clusters.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  yield put(
+    appendEntryToLiveFeed({
```
[the remark this answers is not in the document]
Comment (jamie-suse) [reply assets/js/state/sagas/clusters.js:9]: Okay, will do 👍
Also other sagas in the codebase use `appendEntryToLiveFeed()` though, shouldn't we be removing them also? I can create a tech-debt ticket for this

**Comment by stefanotorresi** [comment]
@jamie-suse I understood this was still undergoing review, why did you self-merge it?
